src/util/expr_cast.h

```diff
@@ -0,0 +1,145 @@
+/*
+  Author: Nathan Phillips <Nathan.Phillips@diffblue.com>
+*/
+
+#ifndef CPROVER_UTIL_EXPR_CAST_H
+#define CPROVER_UTIL_EXPR_CAST_H
+
+/// \file util/expr_cast.h
+/// \brief Templated functions to cast to specific exprt-derived classes
+
+#include <typeinfo>
+#include "invariant.h"
+#include "expr.h"
+
+
+/// \brief Check whether a reference to a generic \ref exprt is of a specific
+///   derived class.
+///
+///   Implement template specializations of this function to enable casting
+///
+/// \tparam T The exprt-derived class to check for
+/// \param base Reference to a generic \ref exprt
+/// \return true if \a base is of type \a T
+template<typename T> bool check_expr_type(const exprt &base);
```

Comment:
Think of a synonym for `type` (`class`? `kind`? Simply `check_expr_cast`?) since the current name suggests we're talking about the unrelated `base.type()`.

Reply:
I like kind fwiw